### PR TITLE
Introduce StateCommitment in StateProviders

### PR DESCRIPTION
**Pull request by frisitano**
Replacement for https://github.com/paradigmxyz/reth/pull/11867

### CODE REVIEW

File: crates/storage/provider/src/providers/state/historical.rs [hunk 1]
```diff
@@ -41,13 +41,15 @@ use std::fmt::Debug;
 /// - [`tables::AccountChangeSets`]
 /// - [`tables::StorageChangeSets`]
 #[derive(Debug)]
-pub struct HistoricalStateProviderRef<'b, Provider> {
+pub struct HistoricalStateProviderRef<'b, Provider, SC> {
```
[the remark this answers is not in the document]
Comment (frisitano) [reply crates/storage/provider/src/providers/state/historical.rs:44]: Great suggestion! This should make things much cleaner. I will implement this pattern now.

[the remark this answers is not in the document]
Comment (frisitano) [reply crates/storage/provider/src/providers/state/historical.rs]: updated.

**Review by klkvr — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
really sorry about the slow progress, 
lots of movement in this codesection, but this is now gradually stabilizing

**Comment by frisitano** [comment]
> really sorry about the slow progress, lots of movement in this codesection, but this is now gradually stabilizing

No problem I understand, thanks for keeping this in mind and pushing it forward 🙏

**Pull request merged**